src/app/fdctl/run/tiles/fd_net.c

```diff
@@ -242,7 +281,77 @@ after_frag( void *             _ctx,
   if( FD_UNLIKELY( route_loopback( ctx->src_ip_addr, *opt_sig ) ) ) {
     ctx->lo_tx->send_func( ctx->xsk_aio[ 1 ], &aio_buf, 1, NULL, 1 );
   } else {
-    ctx->tx->send_func( ctx->xsk_aio[ 0 ], &aio_buf, 1, NULL, 1 );
+    /* extract dst ip */
+    uchar const * p = ctx->frame;
+    uint  ethtype     = ( (uint)p[12UL] << 8u ) | ( (uint)p[13UL] << 0u );
+    uint  dst_ip      = ( (uint)p[30UL] << 24u ) | ( (uint)p[31UL] << 16u ) | ( (uint)p[32UL] << 8u ) | (uint)p[33UL];
+    uint  next_hop    = 0U;
+    uchar dst_mac[6]  = {0};
+    uint  if_idx      = 0;
+
+    if( ethtype == 0x0800 ) {
+
+      /* route the packet */
+      /*
+       * determine the destination:
+       *   same host
+       *   same subnet
+       *   other
+       * determine the next hop
+       *   localhost
+       *   gateway
+       *   subnet local host
+       * determine the mac address of the next hop address
+       *   and the local ipv4 and eth addresses */
+      int rtn = fd_ip_route_ip_addr( dst_mac, &next_hop, &if_idx, ctx->ip, dst_ip );
+      if( FD_UNLIKELY( rtn == FD_IP_PROBE_RQD ) ) {
+        /* another fd_net instance might have already resolved this address
+           so simply try another fetch */
+        fd_ip_arp_fetch( ctx->ip );
+        rtn = fd_ip_route_ip_addr( dst_mac, &next_hop, &if_idx, ctx->ip, dst_ip );
+      }
+
+      long now;
+      switch( rtn ) {
+        case FD_IP_PROBE_RQD:
+          /* TODO possibly buffer some data while waiting for ARPs to complete */
+          /* TODO rate limit ARPs */
+          /* TODO add caching of ip_dst -> routing info */
+          send_arp_probe( ctx, next_hop, if_idx );
+
+          /* refresh tables */
+          now = fd_log_wallclock();
+          ctx->ip_next_upd = now + (long)200e3;
+          break;
+        case FD_IP_NO_ROUTE:
+          /* cannot make progress here */
+          break;
+        case FD_IP_SUCCESS:
+          /* set destination mac address */
+          memcpy( ctx->frame, dst_mac, 6UL );
+
+          /* set source mac address */
+          memcpy( ctx->frame + 6UL, ctx->src_mac_addr, 6UL );
+
+          ctx->tx->send_func( ctx->xsk_aio[ 0 ], &aio_buf, 1, NULL, 1 );
```

Comment:
Do you need to recalculate the checksum?

Reply:
the IP checksum doesn't include the ETH header